pkg/sampling/encoding_test.go

```diff
@@ -0,0 +1,304 @@
+
+	// 0x55555554p-32 is very close to 1/3
+	require.InEpsilon(t, 1-1/3., must(tValueToProbability("55555554")), 1e-9)
+}
+
+func TestProbabilityToThreshold(t *testing.T) {
+	require.Equal(t,
+		must(TValueToThreshold("8")),
+		must(ProbabilityToThreshold(0.5)))
+	require.Equal(t,
+		must(TValueToThreshold("ffffffffffffff")),
+		must(ProbabilityToThreshold(0x1p-56)))
+	require.Equal(t,
+		must(TValueToThreshold("ffffffffffff00")),
+		must(ProbabilityToThreshold(0x100p-56)))
+	require.Equal(t,
+		must(TValueToThreshold("00000000000010")),
+		must(ProbabilityToThreshold(1.0-0x1p-52)))
+	require.Equal(t,
+		AlwaysSampleThreshold,
+		must(ProbabilityToThreshold(1.0)))
+
+	zt, err := ProbabilityToThreshold(0)
+	require.Equal(t, zt, AlwaysSampleThreshold)
+	require.Error(t, err)
+	require.Equal(t, err, ErrProbabilityRange)
+}
+
+func TestShouldSample(t *testing.T) {
+	// Test four boundary conditions for 50% sampling,
+	thresh := must(ProbabilityToThreshold(0.5))
+	// Smallest TraceID that should NOT sample.
+	require.False(t, thresh.ShouldSample(TraceIDToRandomness(pcommon.TraceID{
+		// 9 meaningless bytes
+		0xee, 0xee, 0xee, 0xee, 0xee, 0xee, 0xee, 0xee, 0xee,
+		0, // randomness starts here
+		0, 0, 0, 0, 0, 0,
+	})))
+	// Largest TraceID that should NOT sample.
+	require.False(t, thresh.ShouldSample(TraceIDToRandomness(pcommon.TraceID{
+		// 9 meaningless bytes
+		0xee, 0xee, 0xee, 0xee, 0xee, 0xee, 0xee, 0xee, 0xee,
+		0x7f, // randomness starts here
+		0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+	})))
+	// Smallest TraceID that should sample.
+	require.True(t, thresh.ShouldSample(TraceIDToRandomness(pcommon.TraceID{
+		// 9 meaningless bytes
+		0xee, 0xee, 0xee, 0xee, 0xee, 0xee, 0xee, 0xee, 0xee,
+		0x80, // randomness starts here
+		0, 0, 0, 0, 0, 0,
+	})))
+	// Largest TraceID that should sample.
+	require.True(t, thresh.ShouldSample(TraceIDToRandomness(pcommon.TraceID{
+		// 9 meaningless bytes
+		0xee, 0xee, 0xee, 0xee, 0xee, 0xee, 0xee, 0xee, 0xee,
+		0xff, // randomness starts here
+		0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+	})))
+}
+
+func TestRValueSyntax(t *testing.T) {
+	type testCase struct {
+		in        string
+		expectErr error
+	}
+	for _, test := range []testCase{
+		// correct cases
+		{"12341234123412", nil},
+
+		// wrong size
+		{"123412341234120", ErrRValueSize},
+		{"1234123412341", ErrRValueSize},
+		{"", ErrRValueSize},
+
+		// bad syntax
+		{"abcdefgabcdefg", strconv.ErrSyntax},
+	} {
+		t.Run(testName(test.in), func(t *testing.T) {
+			rnd, err := RValueToRandomness(test.in)
+
+			if test.expectErr != nil {
+				require.True(t, errors.Is(err, test.expectErr),
+					"%q: not expecting %v wanted %v", test.in, err, test.expectErr,
+				)
+				require.Equal(t, must(RValueToRandomness("00000000000000")), rnd)
+			} else {
+				require.NoError(t, err, "%q", test.in)
+
+				val, err := strconv.ParseUint(test.in, 16, 64)
+				require.NoError(t, err)
+
+				require.Equal(t, TraceIDToRandomness(
+					pcommon.TraceID{
```

Comment:
```suggestion
					// This explicitly constructs a TraceID from 9 random
					// bytes plus the 7 lowest bytes of the input value.
					pcommon.TraceID{
```
To save future-me from having to decode this again.